### PR TITLE
KEYCLOAK-9360 Two factor authentication with W3C Web Authentication - WIP 1st impl phase

### PR DESCRIPTION
**Pull request by tnorimat**
This documentation PR is for https://github.com/keycloak/keycloak/pull/6248 .

### CODE REVIEW

File: server_admin/topics/authentication/webauthn.adoc
Comment (mposolda) [inline server_admin/topics/authentication/webauthn.adoc:48]: I would add note here, that it is also possible to use value REQUIRED for execution's requirement. In that case, if existing user without WebAuthn credential attached is trying to authenticate, the "WebAuthn Register" required action will be automatically added to his account and he will be required to register webAuthn credential before authentication.

I am personally using this for my manual testing as it doesn't require admin to manually add requiredAction to existing user.

File: server_admin/topics/authentication/webauthn.adoc
Comment (mposolda) [inline server_admin/topics/authentication/webauthn.adoc]: This is true just in case that OPTIONAL execution requirement is chosen for the Authenticator. In case it is REQUIRED, there is no need for admin to manually add required action toexisting  user's accounts. I would possibly also add note about this here?

**Review by mposolda — COMMENTED** [review]
Thanks for the documentation! IMO it's also great that you added info that webauthn support is still in development phase.

I've added 2 notes inline with regards to clarify that authenticator execution requirement can be marked as REQUIRED and then won't require administrator to manually add requiredAction to all existing user accounts.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc:48]: I see. I would like to change "OPTIONAL" to "REQUIRED" here. Is that OK?

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Corresponding to the comment just above (OPTIONAL -> REQUIRED for execution's requirement), What do you think about the following (replacing line 67) ?

User with their account::

When such the users try to log in, they are required to register their WebAuthn authenticator automatically.

File: server_admin/topics/authentication/webauthn.adoc
Comment (mposolda) [reply server_admin/topics/authentication/webauthn.adoc:48]: Yes, that is fine IMO.

File: server_admin/topics/authentication/webauthn.adoc
Comment (mposolda) [reply server_admin/topics/authentication/webauthn.adoc]: +1

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Suggestion:
IMPORTANT: Please note that WebAuthn support is still in development and not yet complete, so we recommend that you use this feature experimentally.  Also, this support's specification and user interfaces may change.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: suggest:  ... supports THE WebAuthn specification.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Since these bullets are complete sentences, please add a period (FULL STOP) at the end of each one.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: ... THE 'Admin Console'

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Suggest:  Set 'User Registration' to ON and click 'Save'.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Please begin each step with a capital letter.... Open 'Realm Settings'...

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: ...THE 'Admin Console'

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Suggest:  Open the...

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: General format suggestions:

- Being with a capital letter:  Push

- Add THE before the noun:   ...the 'Register'...

- Complete full sentences with a period:  button.

So "Push the 'Register ' button.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: THE 'Admin Console'

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: ...THE 'User Password Form'..

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Suggestion:
The appropriate method to register a WebAuthen authenticator depends on if the user has or has not already registered an account on {project_name}.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc:53]: Not sure what this sentence means. Also, it ends with two colons...  "account::"

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Add THE before "Register' link and 'Register' button...

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: THE 'Admin Console'...

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Click Save. instead of "push 'Save' Button."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Fill in items and click Save.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: .. and click Login.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Instead of "is shown" try "appears' -- more active voice
also, please use this format "Click <button-name>"   So it would be "Click Authenticate.'
Finally, I am not sure what "as it is" means after 'Authenticate' button

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: on THE 'Admin Console'

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Following the same format as earlier this would be:
Open the `Users -> (user) -> Credentials` tab.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Click Save.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: instead of "are shown just below" please use simply "follow."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Consider:  This is a mandatory configuration, which is applied to the operation of registering the WebAuthn authenticator.

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try:  For more details, see https://www.w3.org/TR/webauthn/#dictionary-pkcredentialentity[WebAuthn Specification].

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try "signature algorithms to use"...

Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: instead of "that can be used to sign and verify"  try "for signing and verifying"

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try "For more details, see https://www.w3.org/TR/webauthn/#dictdef-publickeycredentialparameters[WebAuthn Specification]."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: This is THE ID as...

Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try "This is an optional configuration item that is applied to the operation of registering WebAuthn authenticator."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try:  "For more details, see  the https://www.w3.org/TR/webauthn/#rp-id[WebAuthn Specification]."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Please add "the" before Attestation Statement"

Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: "This is an optional configuration item that is applied to the operation of registering WebAuthn authenticator."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try:  For more details, see in the https://www.w3.org/TR/webauthn/#attestation-convey[WebAuthn Specification].

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: "This is an optional configuration item that is applied to the operation of registering WebAuthn authenticator."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try:  For more details, see the https://www.w3.org/TR/webauthn/#enumdef-authenticatorattachment[WebAuthn Specification].

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: "This is an optional configuration item that is applied to the operation of registering WebAuthn authenticator."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: For more details, see  the https://www.w3.org/TR/webauthn/#dom-authenticatorselectioncriteria-requireresidentkey[WebAuthn Specification].

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: "This is an optional configuration item that is applied to the operation of registering WebAuthn authenticator and authenticating the user by WebAuthn authenticator."

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try:  For more details, see https:......

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: Try: "For more details, see...

File: server_admin/topics/authentication/webauthn.adoc
Comment (andymunro) [inline server_admin/topics/authentication/webauthn.adoc]: There are two NOs and one NOT in this sentence. "No WebAuthn authenticator whose AAGUID is not on this list can not be registered."  I am struggling to understand this.

I wonder if you remove this sentence  if the paragraph is more clear.

I think the last sentence makes it very clear. "If no entry is set on this list, any WebAuthn authenticator can be registered."

**Review by andymunro — COMMENTED** [review]
Lots of good content.  I have added individual comments, but here are some general rules.

- Use "Click <button-name>"  instead of Push the <button-name> button.

- End bullets that are a full sentence with a period.
- Start steps with a capital letter.
- Remember to use "the" before nouns

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll add a period at the end of each bullet as you commented..

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc:53]: Yes, it seems that this sentence is not needed necessarily. I'll delete it.
"::" stands for the label in adoc.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested. I'll delete "as it is" because it is not needed necessarily.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Thank you. I'll fix this sentence as you suggested.

File: server_admin/topics/authentication/webauthn.adoc
Comment (tnorimat) [reply server_admin/topics/authentication/webauthn.adoc]: Yes, I'll remove the sentence "No WebAuthn authenticator whose AAGUID is not on this list can not be registered." because the remaining other sentences can tell what I want to say.

**Comment by tnorimat** [comment]
@andymunro @mposolda 
Thank you for your comment. I've fixed the sentences by incorporating your comments. Could you please re-check the fixed sentences https://github.com/keycloak/keycloak-documentation/pull/737/commits/423fc4d1342d809c9cf6f5433877e2929c7a18d7 ?

**Review by mposolda — DISMISSED** [review]
@tnorimat Thanks for the changes! Looks good to me know. I think it can be merged once we merge the implementation PR.

@andymunro Thanks for the review of this PR!

**Comment by tnorimat** [comment]
@andymunro 
Thank you. I've fixed by incorporating your suggestions.

@mposolda 
It seems that the latest commit has overridden the previous commit which had incorporated your first time review's comment, which has diminished your approval.

And, I think it might be better for me to squash all commits onto 1 commit as usual. What do yo think about it?

**Review by mposolda — DISMISSED** [review]
Approving, thanks @andymunro for the review and thanks @tnorimat for incorporating changes. You can squash commits, although it is not strictly needed as we have the option to squash the commits during merge now. This is ready to go, once we have implementation PR ready.

**Review by mposolda — APPROVED** [review]

**Pull request merged**
